zerver/lib/custom_zulip_transaction.py

```diff
@@ -0,0 +1,187 @@
+    Although there is no explicit condition (WHERE), nothing actually updates
+    if insertion_cte returns null.
+    """
+    with connection.cursor() as cursor:
+        query = SQL(
+            """
+            WITH status_cte AS MATERIALIZED (
+                SELECT * FROM {idempotency_table}
+                WHERE idempotency_key = %(idempotency_key_value)s
+                FOR UPDATE NOWAIT
+            ),
+            insertion_cte AS(
+                INSERT INTO {dest_table} ({cols})
+                SELECT {values_placeholders}
+                WHERE (SELECT identifier FROM status_cte) IS NULL
+                RETURNING {pk} AS new_id
+            ),
+            update_cte AS(
+                UPDATE {idempotency_table}
+                SET identifier = insertion_cte.new_id
+                FROM insertion_cte
+                WHERE idempotency_key = %(idempotency_key_value)s
+            )
+            SELECT
+                CASE
+                    WHEN (SELECT identifier FROM status_cte) IS NULL THEN 'NEW' ELSE 'DUPLICATE'
+                END AS status,
+                (SELECT new_id FROM insertion_cte),
+                (SELECT cached_response FROM status_cte);
+        """
+        ).format(
+            dest_table=Identifier(destination_model._meta.db_table),
+            idempotency_table=Identifier(IdempotentRequest._meta.db_table),
+            cols=SQL(", ").join(map(Identifier, field_names)),
+            values_placeholders=SQL(", ").join(map(Placeholder, field_names)),
+            pk=Identifier(destination_model._meta.pk.name),
+        )
+        cursor.execute(query, {**column_to_value, "idempotency_key_value": idempotency_key})
+        result = cursor.fetchall()
+        return result
+
+
+# Currently, the decorated function do_work always returns [SentMessageResult],
+# but zulip_transaction decorator is designed to be generic for future support of other work.
+# That's why we check the work being done (by func name) and handle each type accordingly.
+def post_do_work(work_result: WrappedReturnT, work_name: str, idempotency_key: str) -> None:
+    from zerver.actions.message_send import SentMessageResult
+
+    if work_name == "do_send_messages":
+        work_result = work_result[0]  # type: ignore[index] # It is certainly list[SentMessageResult]
+        assert isinstance(work_result, SentMessageResult)
+        # Cache response if it's a new message,
+        # since automatic_new_visibility_policy (conditionally included in a response)
+        # isn't known until do_send_message() completes.
+        if work_result.status == "NEW":
+            sent_messate_result_dict = asdict(work_result)
+            # "status" is only SentMessageResult related.
+            sent_messate_result_dict.pop("status")
+            IdempotentRequest.objects.filter(idempotency_key=idempotency_key).update(
+                cached_response=sent_messate_result_dict
+            )
+
+
+def zulip_transaction(
```

Comment:
I think you want a name like `idempotent_request_transaction` that clarifies what's special about this transaction in the name.